sda/overlay.go

```diff
@@ -30,20 +31,132 @@ type Overlay struct {
 	instancesLock     sync.Mutex
 	protocolInstances map[TokenID]ProtocolInstance
 
+	// treeMarshal that needs to be converted to Tree but host does not have the
+	// entityList associated yet.
+	// map from Roster.ID => trees that use this entity list
+	pendingTreeMarshal map[RosterID][]*TreeMarshal
+	// lock associated with pending TreeMarshal
+	pendingTreeLock sync.Mutex
+
+	// pendingSDAData are a list of message we received that does not correspond
+	// to any local Tree or/and Roster. We first request theses so we can
+	// instantiate properly protocolInstance that will use these SDAData msg.
+	pendingSDAs []*ProtocolMsg
+	// lock associated with pending SDAdata
+	pendingSDAsLock sync.Mutex
+
 	transmitMux sync.Mutex
 }
 
 // NewOverlay creates a new overlay-structure
 func NewOverlay(h *Host) *Overlay {
-	return &Overlay{
-		host:              h,
-		trees:             make(map[TreeID]*Tree),
-		entityLists:       make(map[RosterID]*Roster),
-		cache:             NewTreeNodeCache(),
-		instances:         make(map[TokenID]*TreeNodeInstance),
-		instancesInfo:     make(map[TokenID]bool),
-		protocolInstances: make(map[TokenID]ProtocolInstance),
-		transmitMux:       sync.Mutex{},
+	o := &Overlay{
+		host:               h,
+		trees:              make(map[TreeID]*Tree),
+		entityLists:        make(map[RosterID]*Roster),
+		cache:              NewTreeNodeCache(),
+		instances:          make(map[TokenID]*TreeNodeInstance),
+		instancesInfo:      make(map[TokenID]bool),
+		protocolInstances:  make(map[TokenID]ProtocolInstance),
+		transmitMux:        sync.Mutex{},
+		pendingTreeMarshal: make(map[RosterID][]*TreeMarshal),
+		pendingSDAs:        make([]*ProtocolMsg, 0),
```

Comment:
It actually checks if there are some messages going to protocols where the overlay did not have the tree before and now it has so it can create a `TreeNodeInstance` + `ProtocolInstance` and pass down the message. I changed it to `checkPendingProtocolMessages` , is that all right for you ?